yarn-project/pxe/src/note_processor/utils/add_nullable_field_to_payload.ts

```diff
@@ -0,0 +1,67 @@
+import { L1NotePayload, Note } from '@aztec/circuit-types';
+import { type Fr } from '@aztec/foundation/fields';
+import { ContractNotFoundError } from '@aztec/simulator';
+
+import { type PxeDatabase } from '../../database/pxe_database.js';
+
+/**
+ * Inserts publicly delivered nullable fields into the note payload.
+ * @param db - PXE database used to fetch contract instance and artifact.
+ * @param payload - Note payload to which nullable fields should be added.
+ * @param nullableFields - List of nullable fields to be added to the note payload.
+ * @returns Note payload with nullable fields added.
+ */
+export async function addNullableFieldsToPayload(
+  db: PxeDatabase,
+  payload: L1NotePayload,
+  nullableFields: Fr[],
+): Promise<L1NotePayload> {
+  const instance = await db.getContractInstance(payload.contractAddress);
+  if (!instance) {
+    throw new ContractNotFoundError(
+      `Could not find instance for ${payload.contractAddress.toString()}. This should never happen here as the partial notes flow should be triggered only for non-deferred notes.`,
+    );
+  }
+
+  const artifact = await db.getContractArtifact(instance.contractClassId);
+  if (!artifact) {
+    throw new Error(
+      `Could not find artifact for contract class ${instance.contractClassId.toString()}. This should never happen here as the partial notes flow should be triggered only for non-deferred notes.`,
+    );
+  }
+
+  const noteFields = Object.values(artifact.notes).find(note => note.id.equals(payload.noteTypeId))?.fields;
+
+  if (!noteFields) {
+    throw new Error(`Could not find note fields for note type ${payload.noteTypeId.toString()}.`);
+  }
+
+  // We sort note fields by index so that we can iterate over them in order.
+  noteFields.sort((a, b) => a.index - b.index);
+
+  // Now we insert the nullable fields into the note based on its indices defined in the ABI.
+  const modifiedNoteItems = [...payload.note.items];
+  let indexInNullable = 0;
+  for (let i = 0; i < noteFields.length; i++) {
+    const noteField = noteFields[i];
+    if (noteField.nullable) {
+      if (i == noteFields.length - 1) {
+        // We are processing the last field so we simply insert the rest of the nullable fields at the end
+        modifiedNoteItems.push(...nullableFields.slice(indexInNullable));
```

Comment:
But there should only be one, right? Because we know there's no more note fields

Reply:
I think I'm confused as to what `noteFields` means

Reply:
There is a naming issue here and in the macros. Field can either mean a struct member (as in amount in TokenNote) or a the type (as in Fr). This is super confusing if you serialize struct members into fields.

Here the last "field" can occupy multiple fields because for example U128 token amount gets serialized to 2 Frs.

I could rename it everywhere in a PR up the stack to struct member if you think it clarifies it.

Reply:
Yeah I asked the Noir team if maybe the struct things should be called members instead. It'd be great if we renamed them internally here.

Reply:
Ok ok, will do tomorrow.